### ISSUE TITLE
buffalo_s detector ships with fixed batch=1 despite PR #1781 fixing the export script

### ISSUE THREAD

**Issue by ceyxasm** [deepinsight/insightface#2878]
PR #1781 fixed scrfd2onnx.py to export SCRFD with dynamic batch dimensions. However, the pre-built buffalo_s model pack available for download still ships with the detector ONNX hardcoded to batch=1. The fix never made it into the distributed weights.       
  
### How to reproduce                                                                                                                 
   
```                                                                                                                         
  from insightface.app import FaceAnalysis
  import numpy as np                                                                                                               
   
  face_app = FaceAnalysis(name='buffalo_s', providers=['CUDAExecutionProvider'])                                                   
  face_app.prepare(ctx_id=0, det_size=(640, 640))                                                                                
                                                                                                                                   
  session = face_app.det_model.session                                                                                           
                                                                                                                                   
  # Check declared shape                                                                                                         
  inp = session.get_inputs()[0]          
  print(inp.shape)  # → [1, 3, 640, 640]  — batch axis is 1, not dynamic                                                           
                                                                                                                                   
  # Try actual batched call                                                                                                        
  batch = np.random.randint(0, 255, (2, 3, 640, 640)).astype(np.float32)                                                           
  session.run([o.name for o in session.get_outputs()], {inp.name: batch})                                                          
```
                                                                                                                              
###  Error:
```                                                                                                                           
  [ONNXRuntimeError] : 2 : INVALID_ARGUMENT : Got invalid dimensions for input: input.1                                            
  index: 0 Got: 2 Expected: 1   
```
                                                                                                                                   
*What I expected*
* Since PR #1781 was merged and the export script now supports dynamic batch, we expected the shipped buffalo_s model to reflect   
  this — either via a re-export after the fix, or a note in the docs that users must re-export themselves from PyTorch weights.

### Questions
1. Was buffalo_s exported before PR #1781 was merged and never updated since?
2. Are the original PyTorch checkpoints for the buffalo_s SCRFD detector publicly available so users can re-export themselves using the fixed script?
3. Is there a plan to re-ship the buffalo packs with dynamic batch support?                                                                                                                                                                               

### Workaround
The community has re-exported SCRFD with correct dynamic axes at alonsorobots/scrfd_320_batched (Hugging Face), but this is at   
  320×320 input vs the 640×640 that buffalo_s uses, which affects small face detection quality: https://huggingface.co/alonsorobots/scrfd_320_batched

**Comment by ceyxasm** [comment]
Additional context: preprocessing and post-processing are also a blocker                                                         
* Even if the shipped model is re-exported with dynamic batch support, users face a second problem: the current detect() wrapper only accepts a single image and handles preprocessing/post-processing internally. To actually use batched inference, users would have to bypass the wrapper entirely and reimplement:                  
  - Preprocessing — resize, BGR→RGB, normalization, and stacking N frames into (N, 3, H, W) manually                               
  - Post-processing — anchor decoding across FPN scales (stride 8, 16, 32), per-frame score filtering, and per-frame NMS (NMS cannot run across frames)                                                                                                        
  
We did something similar for the landmark model by calling session.run() directly with a stacked tensor — but landmark post-processing is trivial (reshape + scale). 
Detection post-processing is significantly more involved.                        

Ideally the fix would include not just a re-exported model but also a batched inference wrapper — something like det_model.detect_batch(images) — so users don't have to reimplement anchor decoding themselves.

> @arvindsaraf

**Comment by ceyxasm** [comment]
### Update — attempted onnx2torch approach (also failing)
                                                                                                                                   
We tried converting det_500m.onnx back to PyTorch using onnx2torch, then re-exporting with dynamic_axes.
                                                                                                                                 
Step 1 — onnx2torch conversion: Works fine. onnx2torch.convert(det_500m.onnx) produces a valid PyTorch module.                   
                                                                                                                                 
Step 2 — re-export to ONNX with dynamic batch: Fails at torch.onnx.export with:                                                  
```  
GuardOnDataDependentSymNode: Could not extract specialized integer from data-dependent expression u0                           
  Caused by: onnx2torch/node_converters/gather.py line 70 — return input_tensor[slice_for_take]                                    
  The Shape → Gather nodes (which compute the batch size at runtime) trip up the dynamo-based exporter.                            
```                                                                                                                              

Step 3 — use the onnx2torch PyTorch model directly (no re-export): Runs for B=2, but the outputs are batch-coupled. Running the  same frame as B=1 vs as frame[0] of B=2 gives a score diff of 0.82 (out of a ~[-5, +5] logit range). Frame results contaminate each other — not just a reshape/slicing issue.                                                                                   
                                                                
                                                                                    
Question for maintainers: Is there a known-correct way to get a dynamic-batch version of det_500m.onnx with buffalo_s weights?    


### EDIT
To rule out any ONNX-level artifact, we converted det_500m.onnx to a native PyTorch module using onnx2torch and ran three comparisons on the same real frame:                                                                                              
                                                                                                                                 
  A vs B — ONNXRuntime B=1 vs onnx2torch B=1 (confirms faithful conversion):                                                       
  out[0]: max_diff=0.000000  OK                                                                                                  
  out[1]: max_diff=0.000001  OK                                                                                                    
  out[2]: max_diff=0.000000  OK                                                                                                  
  ... (all 9 outputs < 1e-4)                                                                                                       
                                                                                                                                 
  B vs C — onnx2torch B=1 vs frame[0] of onnx2torch B=2 [img, img] (batch-independence test):                                      
  out[0] K=12800: max_diff=0.76436  MISMATCH                                                                                       
  out[1] K=3200:  max_diff=0.86294  MISMATCH                                                                                       
  out[3] K=51200: max_diff=3.25097  MISMATCH                                                                                       
  ... (all 9 outputs fail)                                                                                                       
                                                                                                                                   
  The PyTorch model produces bit-exact results vs ONNXRuntime for B=1, but passing the same image twice as a batch produces different scores for frame[0] compared to running it alone. The batch-coupling is inside the model computation itself

> Update: Using onnx2torch hooks to trace intermediate activations, the first diverging operation is Slice_104 — a slice on a shape tensor whose value changes with batch size. This feeds into subsequent Reshape nodes that collapse the batch dimension into the spatial dimension (e.g. (12800,1) for B=1, (25600,1) for B=2), mixing both frames into a single flat tensor  before the sigmoid and all downstream ops. This is why the same image produces different scores depending on what else is in the batch.

**Comment by ceyxasm** [comment]
Additional finding — weight mismatch: 
Compared the first conv layer weight (backbone.stem.0.0.weight, shape (16, 3, 3, 3)) between SCRFD500.pth [(public OneDrive)](https://github.com/deepinsight/insightface/blob/master/detection/scrfd/README.md#pretrained-models) and det_500m.onnx (buffalo_s). 
This layer is unique in both models. max_abs_diff  = 0.332, relative diff = 161% — confirming these are different training runs. The buffalo_s weights are not publicly available in PyTorch format

**Comment by ceyxasm** [comment]
Hey, I was able to reliably solve this issue: https://github.com/ceyxasm/insightface-det-batch-fix
Hence closing
